### PR TITLE
Events for msgscan

### LINKED ISSUE #240: emit a message init event from a port?
msgscan now discovers messages from the "MessageAccepted" event of the ORMP protocol. However, I think msgscan is not only suitable for ORMP, but also for various underlying protocols. Message discovery should not depend on a specific protocol.

### LINKED ISSUE #246: Add msgport message id
returns bytes32 as the message id

### LINKED ISSUE #247: Add events to send and recv functions

### PR DESCRIPTION
**Pull request by hujw77**
- Fix #246 
- Fix #247 
- Fix #240

### CODE REVIEW

**Review by jiguantong — APPROVED** [review]

**Comment by jiguantong** [comment]
@wuminzhe review this.

**Comment by jiguantong** [comment]
msgId for Application

**Review by wuminzhe — APPROVED** [review]

File: src/interfaces/IMessagePort.sol [hunk 1]
```diff
@@ -18,16 +18,22 @@
...
     /// @param toChainId The message destination chain id. <https://eips.ethereum.org/EIPS/eip-155>
     /// @param toDapp The user application contract address which receive the message.
     /// @param message The calldata which encoded by ABI Encoding.
     /// @param params Extend parameters to adapt to different message protocols.
-    function send(uint256 toChainId, address toDapp, bytes calldata message, bytes calldata params) external payable;
+    /// @return msgId Return the ID of message.
+    function send(uint256 toChainId, address toDapp, bytes calldata message, bytes calldata params)
+        external
+        payable
+        returns (bytes32 msgId);
```
Comment (boundless-forest) [inline src/interfaces/IMessagePort.sol:36]: Please update the relavant docs and runnable demo.

File: src/user/Application.sol [hunk 2]
```diff
@@ -22,6 +22,13 @@ abstract contract Application {
         _port = msg.sender;
     }
 
+    function _messageId() internal pure returns (bytes32 _msgDataMessageId) {
+        require(msg.data.length >= 84, "!messageId");
+        assembly {
+            _msgDataMessageId := calldataload(sub(calldatasize(), 84))
+        }
+    }
+
```
Comment (boundless-forest) [inline src/user/Application.sol:31]: And this: https://docs.msgport.xyz/build/interfaces/#application

**Pull request merged**